cs3342_09-master/README.md

```diff
@@ -0,0 +1,191 @@
+
+Adjectives are "lazy" or "smelly"
+
+Nouns are "dog" or "cat"
+
+Verbs are "ate" and "ran"
+
+Adverbs are "slowly" and "noisily"
+
+The following are examples of valid sentences:
+
+* The smelly dog ran.
+* The smelly dog ran slowly.
+* The cat ate noisily.
+
+## Q2.1 (7 points)
+
+Write the BNF (not EBNF) description for this language.
+
+## A2.1
+
+\<sentence> ::= "the" + \<adj> + \<noun> + \<verb> + \<adverb>
+\<adj> ::= \<<empty>> | "lazy" | "smelly"
+\<adverb> ::=  \<<empty>> | "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.2 (5 points)
+
+Write this grammar using EBNF with common extensions
+
+## A2.2
+
+\<sentence> ::= "the" + {\<adj>} + \<noun> + \<verb> + [\<adverb>]
+\<adj> ::= "lazy" | "smelly
+\<adverb> ::= "slowly" | "noisily"
+\<noun> ::=   "dog" | "cat"
+\<verb> ::=   "ate" | "ran"
+
+
+## Q2.3 (6 points)
+
+  Draw a diagram for an FSM which recognizes these sentences. Use EOI as the
+  event that occurs at the end-of-input. Start with a state named S0. Receiving
+  the word "the" in that state will transition to state S1. Name the final state
+  END.
+
+  You can hand draw it and snap a picture, or use a drawing tool. Either way,
+  upload the image, and put the file name in the answer below.
+
+  (Hint: my answer has seven states including the start and end states)
+
+
+## A2.3
+
+Name = A2.3
+
+
+
+## Q2.4 (6 points)
+
+Convert this diagram into a table of the form:
+
+Current state | Next word | Next state
+--------------|-----------|-----------
+    S0        |    the    |     S1
+    S1        |   . . .   |   . . .
+
+(hint: my version has 13 entries. Yours _might_ be different)
+
+## A2.4
+
+Current state | Next word | Next state
+--------------|-----------|-----------
+    S0        |    the    |     S1
+    S1        |   lazy    |     S2
+    S1        |   smelly  |     S2
+    S1        |   "empty" |     S2
+    S2        |   dog     |     S3
+    S2        |   cat     |     S3
+    S3        |   ran     |     S4
+    S3        |   ate     |     S4
+    S4        | slowly    |     S5
+    S4        | noisily    |     S5
+    S5        | EOI       |     END
+
+
+## Q2.5 (12 points)
+
+Translate this table into a programming language of your choice. Then write a
+function that takes a list of words (ending "EOI") and runs the list through the
+state machine. If the state machine cannot find a transition for a word when in
+a given state, return false. If the state machine runs out of words and the
+state is not "END" then return false. Otherwise return true.
+
+Then write some unit tests that exercise your function, making sure that it
+recognizes valid sentences and rejects invalid ones.
+
+The answer should appear in one or more source files in the same directory as
+this file. If I need to do anything more that type a single command to run your
+code, include a script or makefile that will do the job.
+
+## A2.5
+
+ProgLangHomework.cpp
```

Comment:
So, here's the code I accidentally revealed last week:

~~~ cpp
typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> state_table;

const int END = 99;

state_table states = {
  { 0, {{ "the", 1 }}},
	{ 1, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 2, {{ "lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
	{ 3, {{ "ate", 4},  {"ran",    4}}},
	{ 4, {{ "slowly", 5}, { "noisily", 5}, {"EOI", END}}},
	{ 5, {{ "EOI", END }}}
};


int getNextState(int currentState, string nextWord) {
	transitions t = states[currentState];
	if (t.count(nextWord))
		return t[nextWord];
	else
		return -1;
}
~~~

and here's yours

~~~ cpp
typedef int state;
typedef std::map<std::string, state> transitions;
typedef std::map<state, transitions> stateTable;
const int END = 99;
stateTable states = {
    {0, {{"the", 1}}},
    {1, {{"lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
    {2, {{"lazy", 2}, {"smelly", 2}, {"dog", 3}, {"cat", 3}}},
    {3, {{"ate", 4}, {"ran", 4}}},
    {4, {{"slowly", 5}, {"noisily", 5}, {"EOI", END}}},
    {5, {{"EOI", END}}}};
int nextState(int curState, string nextword) {
  transitions t = states[curState];
  if (t.count(nextword))
    return t[nextword];
  else
    return -1;
}
~~~

There's no way you could get from your previous table to this code, which is identical (down to the names) to mine.

# 0